### PR TITLE
Upgrade to php-cs-fixer 3

### PR DESCRIPTION
**Pull request by franmomu**
this should fix https://github.com/sonata-project/dev-kit/pull/1632#pullrequestreview-730162680

POC here: https://github.com/sonata-project/SonataAdminBundle/pull/7434

### CODE REVIEW

**Review by jordisala1991 — APPROVED** [review]

**Comment by VincentLanglet** [comment]
When running `php-cs-fixer` on a project
https://github.com/sonata-project/SonataAdminBundle/blob/4.x/Makefile#L46

Am I wrong or are we using the global installation of php-cs-fixer ?
- It require to update regularly the github actions to bump the versions
- It require for the developer to follow the global installation of php-cs-fixer
- It require for the developer to bump the version regularly of the global installation

Shouldn't we use the same strategy than phpstan/psalm by requiring-dev it in the composer.json and using this version for both the Makefile and the github action ?

**Comment by VincentLanglet** [comment]
> When running `php-cs-fixer` on a project
> https://github.com/sonata-project/SonataAdminBundle/blob/4.x/Makefile#L46
> 
> Am I wrong or are we using the global installation of php-cs-fixer ?
> 
> * It require to update regularly the github actions to bump the versions
> * It require for the developer to follow the global installation of php-cs-fixer
> * It require for the developer to bump the version regularly of the global installation
> 
> Shouldn't we use the same strategy than phpstan/psalm by requiring-dev it in the composer.json and using this version for both the Makefile and the github action ?

For instance, the version should already be updated to 3.0.2 now.

**Comment by jordisala1991** [comment]
+1 to that, maybe it is a good idea to add phpcs fixer and phpunit at the same time, so we have only to touch one time the composer.json

**Comment by VincentLanglet** [comment]
> +1 to that, maybe it is a good idea to add phpcs fixer and phpunit at the same time, so we have only to touch one time the composer.json

I just tried to add phpunit on a project and I see one issue
- The phpunit version is decided by dev-kit https://github.com/sonata-project/dev-kit/blob/master/templates/project/phpunit.xml.dist.twig#L38
So if we add the dependency to the composer.json, it will be handled at two place in the dev-kit and in the bundle.
Is there a way to avoid this ?

**Comment by jordisala1991** [comment]
Yes, remove that line on dev kit , since we wont be installing simple-phpunit

**Comment by VincentLanglet** [comment]
> since we wont be installing simple-phpunit

Not sure about this point, we still need feature like
```
<env name="SYMFONY_DEPRECATIONS_HELPER" value="max[self]=0" />
```

The issue https://github.com/sonata-project/dev-kit/issues/1559 was talking about having both phpunit and simple-phpunit in the dependencies (since it's now kinda recommended). But if it's too complicated, we can just stick with the way we're doing with only simple-phpunit.

**Comment by jordisala1991** [comment]
Read the last point on this : https://symfony.com/doc/current/components/phpunit_bridge.html#usage 

You have to register a listener and thats all

**Comment by VincentLanglet** [comment]
> Read the last point on this : https://symfony.com/doc/current/components/phpunit_bridge.html#usage
> 
> You have to register a listener and thats all

But, in order to register the listener, the bridge still need to be installed.

**Comment by jordisala1991** [comment]
> > Read the last point on this : https://symfony.com/doc/current/components/phpunit_bridge.html#usage
> 
> > 
> 
> > You have to register a listener and thats all
> 
> 
> 
> But, in order to register the listener, the bridge still need to be installed.

Yes, both need to be installed but there is no need to invoke simple-phpunit unless you need to test with multiple phpunit version, which we dont (atm)

**Comment by franmomu** [comment]
I'm fine adding them to the packages, maybe it could be interesting to have a way in `dev-kit` to create PRs as a result of a `composer` command or something like that to not have to create manually all the PRs.

Anyway, even if we add it individually, this PR would ease the transition upgrading the configuration for php-cs-fixer.

**Comment by VincentLanglet** [comment]
> Yes, both need to be installed but there is no need to invoke simple-phpunit unless you need to test with multiple phpunit version, which we dont (atm)

If you know exactly what to do I'm not against a POC PR, then I'll replicate it.

> I'm fine adding them to the packages, maybe it could be interesting to have a way in dev-kit to create PRs as a result of a composer command or something like that to not have to create manually all the PRs.

I don't know how hard it would be to do this, but sure it could be interesting.
Until then, I'll survive doing those PR.

> Anyway, even if we add it individually, this PR would ease the transition upgrading the configuration for php-cs-fixer.

Sure I didn't want to be a blocker

File: .github/workflows/lint.yaml [hunk 1]
```diff
@@ -17,7 +17,7 @@ jobs:
               uses: actions/checkout@v2.3.4
 
             - name: Run PHP-CS-Fixer
-              uses: docker://oskarstark/php-cs-fixer-ga:2.18.1
+              uses: docker://oskarstark/php-cs-fixer-ga:3.0.0
```
Comment (VincentLanglet) [inline .github/workflows/lint.yaml:20]:
```suggestion
              uses: docker://oskarstark/php-cs-fixer-ga:3.0.2
```

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
I don't know why it fail, it's supposed to be released: https://github.com/OskarStark/php-cs-fixer-ga/releases/tag/3.0.2

**Comment by franmomu** [comment]
Looks like the tag in docker https://hub.docker.com/r/oskarstark/php-cs-fixer-ga/tags?page=1&ordering=last_updated is missing, ping @OskarStark

**Comment by OskarStark** [comment]
> Looks like the tag in docker https://hub.docker.com/r/oskarstark/php-cs-fixer-ga/tags?page=1&ordering=last_updated is missing, ping @OskarStark

Yes it is, because docker stopped auto builds feature for non pro (paid) accounts and I currently don't have time to migrate to GitHub action and I don't want to pay 60$ a year 😞 

Anyone open for a PR to build and push a tag on `oskarstark/php-cs-fixer-ga` ? I will provide the credentials as SECRETS and add you as a collaborator, that the workflow runs in the PR

**Comment by franmomu** [comment]
> I don't want to pay 60$ a year 😞

That's more than fair.

> Anyone open for a PR to build and push a tag on `oskarstark/php-cs-fixer-ga` ? I will provide the credentials as SECRETS and add you as a collaborator, that the workflow runs in the PR

I've only used docker just as a provider of services, never tried building or tagging.

I've changed again to use php-cs-fixer version to 3.0.0, since the idea of this PR is to make the upgrade of configuration and files automatically.

**Comment by OskarStark** [comment]
@franmomu I now have a paid plan, you can create a PR for 3.0.2 now, cheers

**Pull request merged**